### DISCUSSION TITLE
How the max_size and compression relates

### DISCUSSION THREAD

**Discussion by silverqx** [ccache/ccache#1243] [Questions & answers] [answered]
Currently, I have set the `max_size` for all compilers like this, first `ccache --clear` and also `--zero-stats` and then I did the initial build, then `ccache --show-compression`, and used this formula `<compessed_data> * 3 + 100`, it looks pretty good and I have 100% cache hits (**all this is without PCH**). 🤯

Note: PCH are rapidly increasing the cache size, for my project it's ~4-5 times. ℹ️

<details>
<summary>The result is 📈</summary>
The first value is with a default compression and the second uncompressed (ccache --show-compression)

```
with PCH:

linux-gcc-qt64   - 780MB (5.0GB)
linux-clang-qt63 - 440MB (1.3GB)
linux-gcc-qt5    - 770MB (4.5GB)

w/o PCH:

clang-cl-qt64    - 135MB (880MB)
msvc2022-qt64    - 190MB (1.5GB)
msvc2019-qt5     - 175MB (1.5GB)
linux-gcc-qt64   - 165MB (700MB)
linux-clang-qt63 - 100MB (680MB)
linux-gcc-qt5    - 115MB (665MB)
linux-clang-qt5  -  90MB (640MB)

MSYS2-clang-qt5  - 140MB (900MB)
MSYS2-clang-qt6  - 140MB (895MB)
MSYS2-gcc-qt5    - 180MB (940MB)
MSYS2-gcc-qt6    - 180MB (940MB)
```

</details>

Except `msvc2022 with qt6`, I have set the `max_size = 700MB` and I had 30 cache misses, so I had to increase it to `1400MB`. 😵‍💫

First, the paragraph in man pages is clear 👍

> After a new compilation result has been written to the cache, ccache will update the size and file number statistics for the subdirectory (one of sixteen) to which the result was written. Then, if the size counter for said subdirectory is greater than max_size / 16 or the file number counter is greater than max_files / 16, automatic cleanup is triggered.

My question is if the `max_size / 16` is compared against compressed data size or uncompressed. 🙋‍♂️

I would normally set it to `2G` and don't care, but I'm hitting GitHub's `10G` [limit](https://github.com/silverqx/TinyORM/actions/caches) 🙃

**Comment by jrosdahl** [comment, marked as answer]
> My question is if the `max_size / 16` is compared against compressed data size or uncompressed.

It's compared against the size of the disk blocks that the compressed files occupy on disk. The _Cache size (GB)_ figure in `ccache -s` is exactly the same as _disk blocks_ in the output of `ccache -x`.

> used this formula `<compessed_data> * 3 + 100`

What does this formula try to calculate?

**Reply by silverqx** [reply]
> It's compared against the size of the disk blocks that the compressed files occupy on disk. The _Cache size (GB)_ figure in `ccache -s` is exactly the same as _disk blocks_ in the output of `ccache -x`.

Thx, this answered my question.  👍

> What does this formula try to calculate?

To have enough space to avoid cache misses and set the `max_size` as low as possible because we have only 10G per project on GitHub.

**Comment by afbjorklund** [comment]
You might also want to investigate using different compression levels, if you are limited by space (rather than by time)

Note: the compression time only affects the cache miss, decompression is about the same for all levels on a cache hit

**Comment by silverqx** [comment]
That's true, a higher compression level can help, but the current compression ratio is impressively high, it's ~85%, so there is not much to compress. Eg. on msvc it's 88% 😮:
```
Total data:          184.8 MB (185.0 MB disk blocks)
Compressed data:     184.8 MB (11.9% of original size)
  Original size:       1.5 GB
  Compression ratio: 8.373 x  (88.1% space savings)
Incompressible data:   0.0 kB
```
